README.md
# NAME
 
SQL::Load - This module is a simple SQL file loader

# VERSION

0.05

# SYNOPSIS

File SQL: users.sql

    -- (find)
    SELECT * FROM users WHERE id = ?;
    
    -- # find-all
    SELECT * FROM users ORDER BY id DESC;
    
    -- [insert]
    INSERT INTO users (name, login, password) VALUES (?, ?, ?);

Code:

    use SQL::Load;
    
    my $sql = SQL::Load->new('/home/user/sql/directory/path');
    
    # load users.sql file
    my $users = $sql->load('users');
     
    print $users->name('find')     # SELECT * FROM users WHERE id = ?;
    print $users->name('find-all') # SELECT * FROM users ORDER BY id DESC;
    print $users->name('insert')   # INSERT INTO users (name, login, password) VALUES (?, ?, ?);
    
    print $sql->load('users#find');     # SELECT * FROM users WHERE id = ?; 
    print $sql->load('users#find-all'); # SELECT * FROM users ORDER BY id DESC;
    print $sql->load('users#insert');   # INSERT INTO users (name, login, password) VALUES (?, ?, ?);    
  
# DESCRIPTION

Intended to separate SQL from Perl code, this module provides some functions that make it easier to do so.

# METHODS

## new

    my $sql_load = SQL::Load->new($path);

Construct a new L<SQL::Load>, passing the folder path is required.
The end param is optional, default returns SQLs with a semicolon in the end, for example using break line:

    my $sql = SQL::Load->new('/home/user/sql/directory/path', "\n");

    my $users = $sql->load('users');
     
    print $users->name('find');     # SELECT * FROM users WHERE id = ?\n
    print $users->name('find-all'); # SELECT * FROM users ORDER BY id DESC\n
    print $users->name('insert');   # INSERT INTO users (name, login, password) VALUES (?, ?, ?)\n

## load

    my $method = $sql_load->load('file_name'); 
    my $method = $sql_load->load('file_name.sql');
    my $method = $sql_load->load('file_name#sql_name'); 
    my $method = $sql_load->load('file_name#sql_at');    
    my $method = $sql_load->load('file_name', 1); # reload to get content directly from the file
    
Load the content in the reference and return an instance of L<SQL::Load::Method>.

## reload

    my $method = $sql_load->reload('file_name'); 
    my $method = $sql_load->reload('file_name.sql');
    my $method = $sql_load->reload('file_name');
    my $method = $sql_load->reload('file_name#sql_name');
    my $method = $sql_load->reload('file_name#sql_at');
    
Reload to get content directly from the file without getting from the tmp from reference.

# FILES SQL
    
## Named SQL

B<For you to name the SQL has 3 ways:>

    -- # NAME
    -- (NAME)
    -- [NAME]
    
These three ways have the same result.

**Examples:**

    -- # find
    SELECT * FROM users WHERE id = ?;
    
    -- # find-all
    SELECT * FROM users ORDER BY id DESC;
    
    -- [FindByEmail]
    SELECT * FROM users WHERE email = ?;   
    
    -- (insert)
    INSERT INTO users (name, email, password) VALUES (?, ?, ?);
    
    -- [delete_by_email]
    DELETE FROM users WHERE email = ?;
    
    -- (UpdatePassword)
    UPDATE users SET password = ? WHERE id = ?;
    
**Important informations:**

Names must be CamelCase, snake_case or kebab-case.

Each SQL statement must end with a semicolon(;).

## SQL in list

Must be separated by semicolon(;)

**Examples:**

    CREATE TABLE users (
        id INT NOT NULL AUTO_INCREMENT,
        name VARCHAR (50) NOT NULL,
        email VARCHAR (100) NOT NULL,
        username VARCHAR (40) NOT NULL,
        password VARCHAR (64) NOT NULL,
        PRIMARY KEY (id)
    );

    CREATE TABLE articles (
        id INT NOT NULL AUTO_INCREMENT,
        user_id INT NOT NULL,
        title VARCHAR (255) NOT NULL,
        data TEXT NOT NULL,
        created TIMESTAMP NOT NULL,
        PRIMARY KEY (id)
    );

    INSERT INTO users (name, email, username, password) 
         VALUES ('John', 'john@email.com', 'john', 'john12345');
         
First is 1, second is 2, and third is 3 in the list position.

# SEE ALSO
 
[SQL::Load::Method](https://metacpan.org/pod/SQL::Load::Method)
 
# AUTHOR
 
Lucas Tiago de Moraes, `lucastiagodemoraes@gmail.com`.
 
# COPYRIGHT AND LICENSE
 
This software is copyright (c) 2022 by Lucas Tiago de Moraes.
 
This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.


Changes
0.06 2023-05-30T11:49:00Z
    - Added end param in method new.

0.05 2023-05-29T16:52:00Z
    - Added end param in method new.

0.04 2022-12-03T09:43:00Z
    - Added an option to get the SQL using 'name' and 'at' from the methods 
      load and reload (file-name#name or file-name#at).

0.03 2022-11-16T00:55:00Z
    - Added method reset in SQL::Load::Method.

0.02 2022-11-08T00:00:00Z
    - Added method reload to get content directly from the file.
    - Making method _get_tmp return a reference of SQL::Load::Method.

0.01 2022-11-06T01:54:00Z
    - First release.


lib/SQL/Load/Util.pm
package SQL::Load::Util;

use strict;
use warnings;
use String::CamelCase qw/
    camelize 
    decamelize
/;
use base qw/Exporter/;
 
our @EXPORT_OK = qw/
    name_list
    parse
    remove_extension
    trim
/;

sub name_list {
    my $name = shift;
    
    my @list;
    
    $name = remove_extension($name);
    
    $name =~ s/-/_/xg;
    push @list, camelize $name;
    
    $name = decamelize $name;
    push @list, $name;
    
    $name =~ s/_/-/xg;
    push @list, $name;
    
    return wantarray ? @list : \@list;    
}

sub parse {
    my $content = shift;
    
    # get all sql by name
    my (@data) = $content =~ /--\s*(?:#|\[|\()\s*([\w-]+)\s*(?:|]|\))\n([^;]+;)/g;
    
    # get all sql without name
    unless (@data) {
        my (@list) = $content =~ /([^;]+;)/g;
        
        my $num = 1;
        for my $sql (@list) {
            push(@data, $num++);
            push(@data, trim($sql));
        }
        
        # if got one or nothing, set content as default
        unless (@data) {            
            push(@data, 'default');
            push(@data, trim($content));
        }        
    }
    
    return wantarray ? @data : \@data;    
}

sub remove_extension {
    my $value = shift;
    
    $value =~ s/\.sql$//i if $value;
    
    return $value;
}

sub trim {
    my $value = shift;
    
    $value =~ s/^\s+|\s+$//g;
    
    return $value;
}

1;


lib/SQL/Load/Method.pm
package SQL::Load::Method;

use strict;
use warnings;
use SQL::Load::Util qw/
    parse
    name_list
/;

sub new {
    my ($class, $content) = @_;  
    
    my ($data, $hash, $keys, $list) = $class->_parse($content);
    
    my $self = {
        _data => $data,
        _hash => $hash,
        _keys => $keys,
        _list => $list,
        _next => 0
    };
    
    return bless $self, $class;
}

sub default {
    my $self = shift;
    
    return $self->{_hash}->{default} if exists $self->{_hash}->{default};
    return $self->{_list}->[0]       if exists $self->{_list}->[0];
    
    return;
}

sub name {
    my ($self, $name) = @_;
    
    my $real = exists $self->{_keys}->{$name} ? $self->{_keys}->{$name} : $name;   
                      
    return $self->{_hash}->{$real} if exists $self->{_hash}->{$real};
    
    return;
}

sub next {
    my $self = shift;
    
    if (defined $self->{_next}) {
        my $next = $self->{_next}++;
        
        return $self->{_list}->[$next] if exists $self->{_list}->[$next];
        
        $self->{_next} = undef;
    }
    
    return;
}

sub at {
    my ($self, $index) = @_;
    
    return $self->{_list}->[$index - 1] if exists $self->{_list}->[$index - 1];
    
    return;
}

sub first {
    my $self = shift;
    
    return $self->{_list}->[0];
}

sub last {
    my $self = shift;
    
    return $self->{_list}->[-1];
}

sub replace {
    my $self = shift;
    
    my %replace = @_;
    
    for my $name (keys %replace) {
        my $value = $replace{$name};
        
        # replace in hash
        for my $key (keys %{$self->{_hash}}) {
            $self->{_hash}->{$key} =~ s/${name}/${value}/g;
        }
        
        # replace in list
        for (my $i = 0; $i < scalar(@{$self->{_list}}); $i++) {
            $self->{_list}->[$i] =~ s/${name}/${value}/g;
        }
    }
    
    return $self;
}

sub reset {
    my $self = shift;
    
    my @data = @{$self->{_data}};
    
    my %hash = @data;
    $self->{_hash} = \%hash;
    
    my @list;
    
    for (my $i = 0; $i < scalar(@data); $i += 2) {        
        push(@list, $data[$i + 1]);
    }  
    
    $self->{_list} = \@list; 
    
    $self->{_next} = 0;
    
    return $self;
}

sub _parse {
    my ($self, $content) = @_;
    
    my @data = parse($content);
    my %hash = @data;
    my %keys;
    my @list;
    
    for (my $i = 0; $i < scalar(@data); $i += 2) {
        my $name = $data[$i];
        my @name_list = name_list($name);
        
        for (@name_list) {
            $keys{$_} = $name;
        }
        
        push(@list, $data[$i + 1]);
    }
    
    return (\@data, \%hash, \%keys, \@list);    
}

1;

__END__

=encoding utf8
 
=head1 NAME
 
SQL::Load::Method

=head1 SYNOPSIS

    my $sql = q{
        -- [find]
        SELECT * FROM users WHERE id = ?;
        
        -- [find-by-email]
        SELECT * FROM users WHERE email = ?;        
    };

    my $method = SQL::Load::Method->new($sql);
    
    # SELECT * FROM users WHERE id = ?;
    print $method->name('find');
    
    # SELECT * FROM users WHERE email = ?;
    print $method->name('find-by-email');
    
=head1 DESCRIPTION

L<SQL::Load::Method> contains useful methods to L<SQL::Load>, 
method load from L<SQL::Load> returns a reference.

=head1 METHODS

=head2 new

    my $method = SQL::Load::Method->new($sql);

construct a new L<SQL::Load::Method>.

=head2 default

    $method->default;
    
returns the first SQL or SQL named as default.

=head2 name

    # -- # name
    # -- [name]
    # -- (name)

    $method->name('name');
    
returns SQL by name.
  
=head2 next

    while (my $sql = $method->next) {
        print $sql;
    }
    
returns the next SQL like an iterator.

=head2 at

    $method->at(1);
    
returns the SQL by position in the list, starting with 1.


=head2 first

    $method->first;
    
returns first the SQL.

=head2 last

    $method->last;
    
returns last the SQL.

=head2 replace

    $method->replace('value', 'new_value')->name('find');
    $method->replace(value => 'new_value')->first;
    $method->replace('value1', 'new_value1', 'value2', 'new_value2')->at(1);
    $method->replace(value1 => 'new_value1', value2 => 'new_value2')->last;
    
replaces values and returns the reference itself.

=head1 SEE ALSO
 
L<SQL::Load>.
 
=head1 AUTHOR
 
Lucas Tiago de Moraes, C<lucastiagodemoraes@gmail.com>.
 
=head1 COPYRIGHT AND LICENSE
 
This software is copyright (c) 2022 by Lucas Tiago de Moraes.
 
This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.
 
=cut


    
    


lib/SQL/Load.pm
package SQL::Load;

use strict;
use warnings;
use Carp;
use SQL::Load::Util qw/
    name_list 
    remove_extension
/;
use SQL::Load::Method;

our $VERSION = '0.04';

sub new {
    my ($class, $path) = @_;
    
    # valid if path exists
    croak "Path not defined!" unless $path;
    croak "The '$path' path does not exist!" unless -d $path;    
    
    my $self = {
        _path => $path,
        _data => {},
        _keys => {}
    };    
    
    return bless $self, $class;
}

sub load {
    my ($self, $file_name, $reload) = @_;
    
    my $name;
    
    if ($file_name =~ /^([\w\-\.\/]+)\#([\w\-]+)$/) {
        $file_name = remove_extension($1);
        $name      = $2; 
    } else {
        $file_name = remove_extension($file_name);
    }
    
    if ($file_name && $file_name =~ /^[\w-]+$/) {
        # if true not get tmp
        unless ($reload) {
            # check if exist the key to get tmp
            my $key = $self->_key($file_name);
            
            # check if tmp exists, if true return
            if ($key) {
                return $name 
                     ? $self->_get_tmp($key)->name($name)
                     : $self->_get_tmp($key);
            }
        }
        
        # get name list
        my $name_list = name_list($file_name);
        
        # get file from name list
        my $file = $self->_find_file($name_list);   
        
        # get content
        my $content = $self->_file_content($file);
        
        # set tmp
        $self->_set_tmp($content, $file, $name_list);
        
        return $name 
             ? SQL::Load::Method->new($content)->name($name)
             : SQL::Load::Method->new($content);
    }
    
    croak "the name '$file_name' is invalid!";
}

sub reload {
    my ($self, $file_name) = @_;
    
    return $self->load($file_name, 1);
}

sub _find_file {
    my ($self, $name_list) = @_;
    
    my $file;
    
    for my $name (@$name_list) {
        my $is_file = $self->{_path} . '/' . $name . '.sql';
        
        if (-e $is_file) {
            $file = $is_file;
            
            last;
        }
        
        last if $file;
    }
    
    return $file if $file;
    
    croak "The file does not exist!";    
}

sub _file_content {
    my ($self, $file) = @_;

    my $content = '';
    
    open FH, '<', $file or croak $!;
    while (<FH>) {
       $content .= $_;
    }
    close FH; 
    
    $content =~ s/^\s+|\s+$//g;
    
    return $content;
}

sub _key {
    my ($self, $name) = @_;
    
    return $self->{_keys}->{$name} if exists $self->{_keys}->{$name};
    
    return;
}

sub _generate_key {
    my @characters = ('0'..'9', 'A'..'Z', 'a'..'z');
    my $x = int scalar @characters;
    my $result = join '', map $characters[rand $x], 1..16;

    return $result;
}

sub _get_tmp {
    my ($self, $key) = @_;
    
    return SQL::Load::Method->new($self->{_data}->{$key}->{content}) 
        if exists $self->{_data}->{$key}->{content};
    
    return;
}

sub _set_tmp {
    my ($self, $content, $file, $name_list) = @_;
    
    # generate new key
    my $key = $self->_generate_key;
    
    # save name => key in tmp keys
    $self->{_keys}->{$_} = $key for @$name_list;
    
    # save data in tmp data
    $self->{_data}->{$key}->{content}   = $content;
}

1;

__END__

=encoding utf8
 
=head1 NAME
 
SQL::Load - This module is a simple SQL file loader

=head1 SYNOPSIS

File SQL: users.sql

    -- (find)
    SELECT * FROM users WHERE id = ?;
    
    -- # find-all
    SELECT * FROM users ORDER BY id DESC;
    
    -- [insert]
    INSERT INTO users (name, login, password) VALUES (?, ?, ?);

Code:

    use SQL::Load;
    
    my $sql = SQL::Load->new('/home/user/sql/directory/path');
    
    # load users.sql file
    my $users = $sql->load('users');
     
    print $users->name('find');     # SELECT * FROM users WHERE id = ?;
    print $users->name('find-all'); # SELECT * FROM users ORDER BY id DESC;
    print $users->name('insert');   # INSERT INTO users (name, login, password) VALUES (?, ?, ?);
    
    print $sql->load('users#find');     # SELECT * FROM users WHERE id = ?; 
    print $sql->load('users#find-all'); # SELECT * FROM users ORDER BY id DESC;
    print $sql->load('users#insert');   # INSERT INTO users (name, login, password) VALUES (?, ?, ?);
  
=head1 DESCRIPTION

Intended to separate SQL from Perl code, this module provides some functions that make it easier to do so.

=head1 METHODS

=head2 new

    my $sql_load = SQL::Load->new($path);

Construct a new L<SQL::Load>, passing the folder path is required.

=head2 load

    my $method = $sql_load->load('file_name'); 
    my $method = $sql_load->load('file_name.sql');
    my $method = $sql_load->load('file_name#sql_name'); 
    my $method = $sql_load->load('file_name#sql_at');
    my $method = $sql_load->load('file_name', 1); # reload to get content directly from the file
    
    
Load the content in the reference and return an instance of L<SQL::Load::Method>.

=head2 reload

    my $method = $sql_load->reload('file_name'); 
    my $method = $sql_load->reload('file_name.sql');
    my $method = $sql_load->reload('file_name');
    my $method = $sql_load->reload('file_name#sql_name');
    my $method = $sql_load->reload('file_name#sql_at');
    
Reload to get content directly from the file without getting from the tmp from reference.

=head1 FILES SQL
    
=head2 Named SQL

B<For you to name the SQL has 3 ways:>

    -- # NAME
    -- (NAME)
    -- [NAME]
    
These three ways have the same result.

B<Examples:>

    -- # find
    SELECT * FROM users WHERE id = ?;
    
    -- # find-all
    SELECT * FROM users ORDER BY id DESC;
    
    -- [FindByEmail]
    SELECT * FROM users WHERE email = ?;   
    
    -- (insert)
    INSERT INTO users (name, email, password) VALUES (?, ?, ?);
    
    -- [delete_by_email]
    DELETE FROM users WHERE email = ?;
    
    -- (UpdatePassword)
    UPDATE users SET password = ? WHERE id = ?;
    
B<Important informations:>

Names must be CamelCase, snake_case or kebab-case.

Each SQL statement must end with a semicolon(;).

=head2 SQL in list

Must be separated by semicolon(;)

B<Examples:>

    CREATE TABLE users (
        id INT NOT NULL AUTO_INCREMENT,
        name VARCHAR (50) NOT NULL,
        email VARCHAR (100) NOT NULL,
        username VARCHAR (40) NOT NULL,
        password VARCHAR (64) NOT NULL,
        PRIMARY KEY (id)
    );

    CREATE TABLE articles (
        id INT NOT NULL AUTO_INCREMENT,
        user_id INT NOT NULL,
        title VARCHAR (255) NOT NULL,
        data TEXT NOT NULL,
        created TIMESTAMP NOT NULL,
        PRIMARY KEY (id)
    );

    INSERT INTO users (name, email, username, password) 
         VALUES ('John', 'john@email.com', 'john', 'john12345');
         
First is 1, second is 2, and third is 3 in the list position.

=head1 SEE ALSO
 
L<SQL::Load::Method>.
 
=head1 AUTHOR
 
Lucas Tiago de Moraes, C<lucastiagodemoraes@gmail.com>.
 
=head1 COPYRIGHT AND LICENSE
 
This software is copyright (c) 2022 by Lucas Tiago de Moraes.
 
This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.
 
=cut
